AndroidSDKCore/src/main/java/com/leanplum/internal/Util.java

```diff
@@ -194,23 +197,40 @@ private static String getWifiMacAddressHash(Context context) {
   }
 
   /**
-   * Retrieves the advertising ID. Requires Google Play Services. Note: This method must not run on
-   * the main thread.
+   * Retrieves the advertising ID. Requires Google Play Services or androidX. Note: This method must
+   * not run on the main thread.
    */
-  private static DeviceIdInfo getAdvertisingId(Context caller) throws Exception {
+  private static DeviceIdInfo getAdvertisingId(Context caller) {
     try {
-      // Using reflection because the app will either crash or print warnings
-      // if the app doesn't link to Google Play Services, even if this method is not called.
-      Object adInfo = Class.forName("com.google.android.gms.ads.identifier.AdvertisingIdClient")
-          .getMethod("getAdvertisingIdInfo", Context.class).invoke(null, caller);
-      String id = checkDeviceId(
-          "advertising id", (String) adInfo.getClass().getMethod("getId")
+      final String[] classNames = {
+          "androidx.ads.identifier.AdvertisingIdClient",
+          "com.google.android.gms.ads.identifier.AdvertisingIdClient"
+      };
+
+      for (String name : classNames) {
+        try {
+          Object adInfo = Class.forName(name)
+              .getMethod("getAdvertisingIdInfo", Context.class)
+              .invoke(null, caller);
+
+          if (name.equals(classNames[0])) {
+            Method get = adInfo.getClass().getMethod("get", long.class, TimeUnit.class);
+            adInfo = get.invoke(adInfo, 5, TimeUnit.SECONDS);
```

Comment:
why do we need to do this timeout? Is this us handling the exception instead of passing that over?

Reply:
The return type is future, need to wait for few seconds to get the result back